R/data_stitch.R

```diff
@@ -0,0 +1,116 @@
+  # 1. Test function arguments
+  #-----------------------------------------
+
+  # Test that x is of class sf
+  assertthat::assert_that(inherits(x, "sf"),
+                          msg = paste( "x is not an sf dataframe."
+                          )
+  )
+
+
+  #Test that contained_taxa is of class numeric
+  assertthat::assert_that(is.numeric(to_stitch),
+                          msg = paste( "to_stitch is not of class numeric."
+                          )
+  )
+
+  #Test that taxa_complex_name is of class character
+  assertthat::assert_that(is.character(taxa_complex_name),
+                          msg = paste( "taxa_complex_name is not of class character."
+                          )
+  )
+
+  #Test that, when provided, taxa_complex_key is of class integer
+  if (!is.null(taxa_complex_key)) {
+    assertthat::assert_that(is.numeric(taxa_complex_key),
+                            msg = "taxa_complex_key should be of class numeric."
+    )
+  }
+
+  #Test that all necessary columns are present
+  columnnames<-c("acceptedTaxonKey",
+                 "decimalLatitude",
+                 "decimalLongitude",
+                 "coordinateUncertaintyInMeters",
+                 "acceptedScientificName",
+                 "year_cat",
+                 "n_obs",
+                 "geometry")
+  colnames_x <-names(x)
+
+  missing_in_x<- setdiff(columnnames, colnames_x)
+
+
+  assertthat::assert_that(all(columnnames %in% colnames_x),
+                          msg = paste( "The following columns are missing in", deparse(substitute(x)),":", paste(missing_in_x, collapse = ", "))
+  )
+
+  #-----------------------------------------
+  # 2. Data prep
+  #-----------------------------------------
+
+  #Create a dataset with data that should be stitched and one with data that should not be touched
+  data_to_keep<- x %>%
+    dplyr::filter(!acceptedTaxonKey %in% to_stitch)
+
+  data_to_stitch<- x %>%
+    dplyr::filter(acceptedTaxonKey %in% to_stitch)
+
+  # Add a common taxonkey and scientific name to the data_to_stich dataset
+  data_to_stitch$acceptedScientificName<-taxa_complex_name
+
+  if(!is.null (taxa_complex_key)) {
+  data_to_stitch$acceptedTaxonKey<- taxa_complex_key
+  }else{
+  data_to_stitch$acceptedTaxonKey<- dplyr::first(unique(data_to_stitch$acceptedTaxonKey))
+  }
+
+  #Recalculate the n_obs per year_category for this species complex
+  data_to_stitch<-data_to_stitch %>%
+    dplyr::group_by(.data$year_cat,
+                    .data$acceptedScientificName,
+                    .data$acceptedTaxonKey,
+                    .data$decimalLatitude,
+                    .data$decimalLongitude) %>%
+    dplyr::mutate(coordinateUncertaintyInMeters = ifelse(all(is.na(.data$coordinateUncertaintyInMeters)),
+                                                         NA_real_,
+                                                         min(.data$coordinateUncertaintyInMeters, na.rm = TRUE)))%>%
+    dplyr::summarize(n_obs = dplyr::n(),
```

Comment:
in case of the example I get the following:
```
> sum(df$n_obs)
[1] 17329
> sum(df_stitched$n_obs)
[1] 13657
```

Reply:
Thanks for noticing! I think I am most fan of adding logic to data_prep to make sure n is always 1. I was gonna add this to climate_cast, where one row should be equal to one observation but it makes more sense in data_prep because we would want to have more than 1 observation when it comes to a species complex (in the case that we have 1 observation of species 1 and 1 observation of species 2 this should count as 2 observations of the complex for climate cast). I'll create an issue to add this logic to data_prep().

Reply:
If n_obs is always 1 in data_prep() I don't think this code needs changes, as it will sum all rows for each group.

Reply:
Indeed no changes needed if data_prep returns 1 row = 1 n_obs.